### PR TITLE
ChatLog - Luqi Xie C22 Pheonix

### PR DESCRIPTION
**Pull request by shiqipaper**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,17 +3,21 @@
...
-    "test": "vitest --run"
+    "test": "vitest --run",
+    "deploy": "gh-pages -d dist"
   },
   "dependencies": {
     "luxon": "^2.5.2",
     "react": "^18.3.1",
-    "react-dom": "^18.3.1"
+    "react-dom": "^18.3.1",
+    "predeploy": "npm run build",
```
Comment (anselrognlie) [inline package.json:19]: 👀 This line `"predeploy"` is intended to define a script, but is listed in the dependencies section. This prevented npm from being able to install the project. This should be moved up to the `"scripts"` section.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,30 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+
+
+const ChatEntry = ({ id, sender, body, timeStamp, liked, onLike }) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:6]: 👍 I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. We do need to remember that these are all passed in as a single object (the one we usually call `props`) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params) I really prefer the glanceability.

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,30 @@
...
   return (
     <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+      <h2 className="entry-name">{sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time"><TimeStamp time={timeStamp}></TimeStamp></p>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:14]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

Note that the `TimeStamp` component doesn't receive children components (it resembles "void" HTML tags such as `input` or `img` rather than tags like `p` or `a` that can have children). For such components, prefer to write them using "self-closing" style (note the `/>` to complete the tag rather than a regular `>`).
```html
<TimeStamp time={timeStamp} />
```

File: src/components/ChatEntry.jsx [hunk 4]
```diff
@@ -1,20 +1,30 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool.isRequired,
+  onLike: PropTypes.func,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:27]: The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`, as you did.

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: dist/assets/index-D8dykwLT.js
Comment (anselrognlie) [inline dist/assets/index-D8dykwLT.js:1]: It should be possible to add `dist` to the `.gitignore` so the `dist` folder doesn't get checked in to your code branch. Build products generally shouldn't get checked into the source repo.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        sender={entry.sender}
+        body={entry.body}
+        timeStamp={entry.timeStamp}
+        liked={entry.liked}
+        onLike={onLike}
+      />);
+  });
+
+  return (
+    < section className="chat-log">
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:20]: Nit: leave no space between the `<` and the tag name.

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import './ChatLog.css';
+import PropTypes from 'prop-types';
+import ChatEntry from './ChatEntry';
+
+const ChatLog = ({ entries, onLike }) => {
+  const chatComponents = entries.map((entry) => {
+    return (
+      <ChatEntry
+        id={entry.id}
+        key={entry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:10]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import './ChatLog.css';
+import PropTypes from 'prop-types';
+import ChatEntry from './ChatEntry';
+
+const ChatLog = ({ entries, onLike }) => {
+  const chatComponents = entries.map((entry) => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:6]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping storing the result into a variable we use in the JSX result as you did here (components are functions, so we can run JS code as usual before we reach the `return`, and even sometimes have multiple `return` statements with different JSX), we could make a helper function that we call as part of the `return`, or this expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component, though can make debugging a little more tricky. But any of those approaches will work fine.

File: package.json [hunk 8]
```diff
@@ -3,17 +3,21 @@
   "private": true,
   "version": "0.0.0",
   "type": "module",
+  "homepage": "https://shiqipaper.github.io/react-chatlog",
```
Comment (anselrognlie) [inline package.json:6]: Since you set the `base` in `vite.config.js` it shouldn't be necessary to set `homepage` here. I'm not sure what effect if any this has for a vite build.

File: src/App.jsx [hunk 9]
```diff
@@ -1,14 +1,39 @@
 import './App.css';
+import ChatLog from './components/ChatLog';
+import CHATDATA from './data/messages.json';
+import { useState } from 'react';
+
 
 const App = () => {
+  const [chatData, setChatData] = useState(CHATDATA);
+
+  const toggleLike = (id) => {
```
Comment (anselrognlie) [inline src/App.jsx:10]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/App.jsx [hunk 10]
```diff
@@ -1,14 +1,39 @@
...
+    setChatData(chatData => chatData.map(chat => {
```
Comment (anselrognlie) [inline src/App.jsx:11]: Nice use of the callback setter style. In this application, it doesn't really matter whether we use the callback style or the value style, but it's good practice to get in the habit of using the callback style.

Comment (anselrognlie) [reply src/App.jsx:11]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 11]
```diff
@@ -1,14 +1,39 @@
...
+      if (chat.id === id) {
+        return {...chat, liked: !chat.liked };
```
Comment (anselrognlie) [inline src/App.jsx:13]: We showed this approach in class, but technically, we're mixing a few responsibilities here. rather than this function needing to know _how_ to change the liked status itself, we could move this update logic to a helper function. This would better mirror how we eventually update records when there's an API call involved.

In this project, our messages are very simple objects, but if we had more involved operations, it could be worthwhile to create an actual class with methods to work with them, or at least have a set of dedicated helper functions to centralize any such mutation logic.

File: src/App.jsx [hunk 12]
```diff
@@ -1,14 +1,39 @@
...
+    }));
+  };
+
+  const calculateTotalLikes = (chatData) => {
+    return chatData.reduce((total, chat) => {
+      return total + chat.liked;
+    }, 0);
+  };
+
+  const totalLikes = calculateTotalLikes(chatData);
```
Comment (anselrognlie) [inline src/App.jsx:26]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

I like that you wrapped your `reduce` call in a well-named helper function. This is one way that we can make using `reduce` a little more understandable for other programmers reading our code, since the syntax can be a little confusing. Another way to make this self-describing is to give a clear name to the function we _pass_ to `reduce`.
```js
const sumLikes = (total, chat) => {
  return chat.liked ? total + 1 : total;
};

const totalLikes = chatData.reduce(sumLikes, 0);

File: src/App.jsx [hunk 13]
```diff
@@ -1,14 +1,39 @@
...
+        return {...chat, liked: !chat.liked };
+      } else {
+        return chat;
+      }
+    }));
+  };
+
+  const calculateTotalLikes = (chatData) => {
+    return chatData.reduce((total, chat) => {
+      return total + chat.liked;
```
Comment (anselrognlie) [inline src/App.jsx:22]: This works as is, since `liked` is a boolean value, and JS will coerce `true` to `1` and `false` to `0`. Still, it can be surprising to see this. Consider using the `liked` boolean value to determine whether to increment the count
```js
      return chat.liked ? total + 1 : total;
```
or to explicitly pick the value to add
```js
      return total + (chat.liked ? 1 : 0)
```

File: src/components/ChatEntry.jsx [hunk 14]
```diff
@@ -1,20 +1,30 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+
+
+const ChatEntry = ({ id, sender, body, timeStamp, liked, onLike }) => {
+  const heartColor = liked ? '❤️' : '🤍';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

File: src/components/ChatEntry.jsx [hunk 15]
```diff
@@ -1,20 +1,30 @@
...
     <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+      <h2 className="entry-name">{sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time"><TimeStamp time={timeStamp}></TimeStamp></p>
+        <button className="like" onClick={() => onLike(id)}>{heartColor}</button>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:15]: 👍 We need a wrapper of some kind rather than calling the received callback through props, since our callback function is expecting a message id as its parameter. If we tried to use it directly as the click event handler, React would end up passing it a clink event, since any function registered as an event handler will always be given the event detail information as its argument.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
